File: hexample-parent/domain/src/main/java/at/willhaben/sp/hexample/domain/recommendation/services/JobRecommendationService.java
package at.willhaben.sp.hexample.domain.recommendation.services;

import at.willhaben.sp.hexample.domain.model.JobOffer;
import at.willhaben.sp.hexample.domain.recommendation.ports.JobRecommendationDatabasePort;
import at.willhaben.sp.hexample.domain.recommendation.ports.JobRecommendationEnginePort;
import at.willhaben.sp.hexample.domain.recommendation.ports.JobRecommendationServicePort;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class JobRecommendationService implements JobRecommendationServicePort {

    private JobRecommendationEnginePort jobRecommendationEngine;
    private JobRecommendationDatabasePort jobRecommendationDatabase;

    public JobRecommendationService(JobRecommendationEnginePort jobRecommendationEngine, JobRecommendationDatabasePort jobRecommendationDatabase) {
        this.jobRecommendationEngine = jobRecommendationEngine;
        this.jobRecommendationDatabase = jobRecommendationDatabase;
    }

    public List<JobOffer> getRecommendationsForUser(Integer userId) {
        List<Integer> recommendedJobsForUser = jobRecommendationEngine.getRecommendedJobIdsForUserId(userId);
        return jobRecommendationDatabase.getJobsForIdList(recommendedJobsForUser);
    }
}

File: hexample-parent/domain/src/main/kotlin/at/willhaben/sp/hexample/domain/recommendation/JobRecommendationService.kt
package at.willhaben.sp.hexample.domain.recommendation

import at.willhaben.sp.hexample.domain.model.JobOffer
import at.willhaben.sp.hexample.domain.recommendation.ports.JobRecommendationDatabasePort
import at.willhaben.sp.hexample.domain.recommendation.ports.JobRecommendationEnginePort
import at.willhaben.sp.hexample.domain.recommendation.ports.JobRecommendationServicePort
import org.springframework.stereotype.Service

@Service
class JobRecommendationService(
    private val jobRecommendationEngine: JobRecommendationEnginePort,
    private val jobRecommendationDatabase: JobRecommendationDatabasePort
) : JobRecommendationServicePort {

    override fun getRecommendationsForUser(userId: Int): List<JobOffer> {
        val recommendedJobsForUser = jobRecommendationEngine.getRecommendedJobIdsForUserId(userId)
        return jobRecommendationDatabase.getJobsForIdList(recommendedJobsForUser)
    }

}
